strawberryfields/tdm/tdmprogram.py

```diff
@@ -303,6 +304,98 @@ def context(self, *args, copies=1, shift="default"):
         self.loop_vars = self.params(*[f"p{i}" for i in range(len(args))])
         return self
 
+    def compile(self, *, device=None, compiler=None):
+        """Compile the time-domain program given a Strawberry Fields photonic hardware device specification.
+        At this stage the compilation is simply a check that the program matches the device.
+
+        Args:
+            device (~strawberryfields.api.DeviceSpec): device specification object to use for
+                program compilation
+            compiler (str, ~strawberryfields.compilers.Compiler): Compiler name or compile strategy
+                to use. If a device is specified, this overrides the compile strategy specified by
+                the hardware :class:`~.DeviceSpec`. For :class:`~.TDMProgram` this should be set to
+                None.
+
+        Returns:
+            Program: compiled program
+        """
+        if compiler is not None:
+            raise CircuitError("Program can only be compiled with default compiler")
```

Comment:
Just curious: would this be applicable should there be multiple tdm compilers? That is, find it a bit strange that a `compiler` can be explicitly passed atm but an error is raised if that does happen

Reply:
@josh146 might know better what is the logic behind this.

Reply:
Not so much logic, more so I was attempting to stick to the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) 😆 I was worried that changing the signature of an inherited method would result in user error elsewhere.

But happy to remove it from the signature if we think that will be less confusing.

Reply:
Fair point, thanks for the clarification!

Reply:
Another idea (which might be completely unreasonable) would be to let the user define any compiler, and then simply call `super().compile(device=device, compiler=compiler)` and hope for the best. One thing that won't pass is the `bb.match_template` function, but that could either be fixed or skipped for `TDMPrograms`. Not sure if there are other things that this will break though. :thinking: